examples/gno.land/p/demo/merkle/merkle.gno

```diff
@@ -0,0 +1,134 @@
+package merkle
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"math"
+)
+
+type Hashable interface {
+	Bytes() []byte
+}
+
+type nodes []Node
+
+type Node struct {
+	hash []byte
+
+	position uint8
+}
+
+func (n Node) Hash() string {
+	return hex.EncodeToString(n.hash[:])
+}
+
+type Tree struct {
+	layers []nodes
+}
+
+func (t *Tree) Root() string {
+	for _, l := range t.layers {
+		if len(l) == 1 {
+			return l[0].Hash()
+		}
+	}
+	return ""
+}
+
+func NewTree(data []Hashable) *Tree {
+	tree := &Tree{}
+
+	leaves := make([]Node, len(data))
+
+	tree.layers = make([]nodes, 1)
+	for i, d := range data {
+		hash := sha256.Sum256(d.Bytes())
+
+		tree.layers[0] = append(tree.layers[0], Node{
+			hash: hash[:],
+		})
+
+		leaves[i] = Node{hash: hash[:]}
+	}
+
+	for len(leaves) > 1 {
+		var level []Node
+		for i := 0; i < len(leaves); i += 2 {
+			var buff bytes.Buffer
+
+			if i < len(leaves)-1 {
+				buff.Write(leaves[i].hash)
+				buff.Write(leaves[i+1].hash)
+				hash := sha256.Sum256(buff.Bytes())
+				level = append(level, Node{
+					hash: hash[:],
+				})
+			} else {
+				level = append(level, leaves[i])
+			}
+		}
+		leaves = level
+		tree.layers = append(tree.layers, level)
+	}
+	return tree
+}
+
+// Proof return a MerkleProof
+func (t *Tree) Proof(data Hashable) ([]Node, error) {
+	targetHash := sha256.Sum256(data.Bytes())
+	targetIndex := -1
+
+	for i, layer := range t.layers[0] {
+		if bytes.Equal(targetHash[:], layer.hash) {
+			targetIndex = i
+			break
+		}
+	}
+
+	if targetIndex == -1 {
+		return nil, errors.New("target not found")
+	}
+
+	var proofs []Node
+
+	for _, layer := range t.layers {
+		var pairIndex int
+
+		if targetIndex%2 == 0 {
+			pairIndex = targetIndex + 1
+		} else {
+			pairIndex = targetIndex - 1
+		}
+		if pairIndex < len(layer) {
+			proofs = append(proofs, Node{
+				hash:     layer[pairIndex].hash,
+				position: uint8(targetIndex) % 2,
+			})
+		}
+		targetIndex = int(math.Floor(float64(targetIndex) / 2))
```

Comment:
```suggestion
		targetIndex /= 2
```

Integer division floors by default, so the passage through float64 is unnecessary and risks being imprecise at very large numbers.